File: web/src/lib/hudCapabilities.ts
export interface HudCapabilitySource {
  available?: boolean;
  enabled?: boolean;
  url?: string;
  directUrl?: string;
  passthroughEnabled?: boolean;
  directEntryEnabled?: boolean;
  reason?: string;
}

export interface HudEntryState {
  available: boolean;
  passthroughEnabled: boolean;
  directEntryEnabled: boolean;
  directUrl: string | null;
  disabledReason: string | null;
}

function normalizeHudUrl(source: HudCapabilitySource | null | undefined): string | null {
  const directUrl = (source?.directUrl ?? source?.url ?? "").trim();
  return directUrl === "" ? null : directUrl;
}

export function resolveHudEntryState(source: HudCapabilitySource | null | undefined): HudEntryState {
  const directUrl = normalizeHudUrl(source);
  const configured = directUrl !== null;
  const backendEnabled = source?.enabled ?? source?.available ?? false;
  const available = backendEnabled && configured;

  return {
    available,
    passthroughEnabled: available && (source?.passthroughEnabled ?? true),
    directEntryEnabled: available && (source?.directEntryEnabled ?? true),
    directUrl,
    disabledReason: available
      ? null
      : (source?.reason ??
        (!configured ? "Loom HUD URL is not configured" : "Loom HUD is disabled")),
  };
}

File: internal/api/handlers/hud_capabilities.go
package handlers

import "net/http"

// HUDCapabilitiesResponse advertises whether FlexDeck can pass through to Loom HUD
// and whether the direct upstream HUD entrypoint is available.
type HUDCapabilitiesResponse struct {
	Available          bool   `json:"available"`
	PassthroughEnabled bool   `json:"passthroughEnabled"`
	DirectEntryEnabled bool   `json:"directEntryEnabled"`
	DirectURL          string `json:"directUrl,omitempty"`
	Reason             string `json:"reason,omitempty"`
}

// HUDCapabilities returns the stable HUD capability contract for the frontend.
func (h *Handler) HUDCapabilities(w http.ResponseWriter, r *http.Request) {
	respondJSON(w, http.StatusOK, h.hudCapabilities())
}

func (h *Handler) hudCapabilities() HUDCapabilitiesResponse {
	directURL := h.loomHUDDirectURL()
	passthroughEnabled := h.loomHUDPassthroughEnabled()
	directEntryEnabled := h.loomHUDDirectEntryEnabled()
	available := passthroughEnabled || directEntryEnabled

	resp := HUDCapabilitiesResponse{
		Available:          available,
		PassthroughEnabled: passthroughEnabled,
		DirectEntryEnabled: directEntryEnabled,
		DirectURL:          directURL,
	}

	if available {
		return resp
	}

	switch {
	case h == nil || h.cfg == nil:
		resp.Reason = "loom hud is unavailable"
	case h.cfg.LoomHUD.Disabled:
		resp.Reason = "loom hud is disabled"
	case h.loomHUDURL() == "" && directURL == "":
		resp.Reason = "loom hud url is not configured"
	default:
		resp.Reason = "loom hud is unavailable"
	}

	return resp
}
